ibc-clients/ics07-tendermint/src/client_state/common.rs

```diff
@@ -0,0 +1,236 @@
+/// in order to make the ClientState APIs more flexible.
+pub fn validate_proof_height(
+    client_state: &ClientStateType,
+    proof_height: Height,
+) -> Result<(), ClientError> {
+    let latest_height = client_state.latest_height;
+
+    if latest_height < proof_height {
+        return Err(ClientError::InvalidProofHeight {
+            latest_height,
+            proof_height,
+        });
+    }
+
+    Ok(())
+}
+
+/// Perform client-specific verifications and check all data in the new
+/// client state to be the same across all valid Tendermint clients for the
+/// new chain.
+///
+/// You can learn more about how to upgrade IBC-connected SDK chains in
+/// [this](https://ibc.cosmos.network/main/ibc/upgrades/quick-guide.html)
+/// guide.
+///
+/// Note that this function is typically implemented as part of the
+/// [`ClientStateCommon`] trait, but has been made a standalone function
+/// in order to make the ClientState APIs more flexible.
+pub fn verify_upgrade_client(
+    client_state: &ClientStateType,
+    upgraded_client_state: Any,
+    upgraded_consensus_state: Any,
+    proof_upgrade_client: CommitmentProofBytes,
+    proof_upgrade_consensus_state: CommitmentProofBytes,
+    root: &CommitmentRoot,
+) -> Result<(), ClientError> {
+    // Make sure that the client type is of Tendermint type `ClientState`
+    let upgraded_tm_client_state = ClientState::try_from(upgraded_client_state.clone())?;
+
+    // Make sure that the consensus type is of Tendermint type `ConsensusState`
+    TmConsensusState::try_from(upgraded_consensus_state.clone())?;
+
+    let latest_height = client_state.latest_height;
+    let upgraded_tm_client_state_height = upgraded_tm_client_state.latest_height();
+
+    // Make sure the latest height of the current client is not greater then
+    // the upgrade height This condition checks both the revision number and
+    // the height
+    if latest_height >= upgraded_tm_client_state_height {
+        Err(UpgradeClientError::LowUpgradeHeight {
+            upgraded_height: latest_height,
+            client_height: upgraded_tm_client_state_height,
+        })?
+    }
+
+    // Check to see if the upgrade path is set
+    let mut upgrade_path = client_state.upgrade_path.clone();
+
+    if upgrade_path.pop().is_none() {
+        return Err(ClientError::ClientSpecific {
+            description: "cannot upgrade client as no upgrade path has been set".to_string(),
+        });
+    };
+
+    let upgrade_path_prefix = CommitmentPrefix::try_from(upgrade_path[0].clone().into_bytes())
+        .map_err(ClientError::InvalidCommitmentProof)?;
+
+    let last_height = latest_height.revision_height();
+
+    // Verify the proof of the upgraded client state
+    verify_membership(
+        client_state,
+        &upgrade_path_prefix,
+        &proof_upgrade_client,
+        root,
+        Path::UpgradeClient(UpgradeClientPath::UpgradedClientState(last_height)),
```

Comment:
according to ibc-go, this should be _plan_ height.

ref: [SetUpgradedClient](https://github.com/cosmos/ibc-go/blob/698cc742b06ab1fc84c93889edd669f4bb949119/modules/core/02-client/keeper/keeper.go#L507).